src/parser_yin.c

```
@@ -0,0 +1,520 @@
+ * @param[out] mod Parsed module structure
+ *
+ * @return LY_ERR values.
+ */
+LY_ERR
+parse_mod(struct lyxml_context *xml_ctx, const char **data, struct lysp_module **mod)
+{
+    LY_ERR ret = LY_SUCCESS;
+    enum yang_keyword kw = YANG_NONE;
+    const char *prefix, *name;
+    size_t prefix_len, name_len;
+    enum yang_module_stmt mod_stmt = Y_MOD_MODULE_HEADER;
+    enum YIN_ARGUMENT arg = YIN_ARG_UNKNOWN;
+
+    char *buf = NULL, *out = NULL;
+    size_t buf_len = 0, out_len = 0;
+    int dynamic;
+
+    /* parse module attributes */
+    while (xml_ctx->status == LYXML_ATTRIBUTE) {
+        ret = lyxml_get_attribute(xml_ctx, data, &prefix, &prefix_len, &name, &name_len);
+        LY_CHECK_ERR_RET(ret != LY_SUCCESS, LOGMEM(xml_ctx->ctx), LY_EMEM);
+
+        arg = match_argument_name(name, name_len);
+
+        switch (arg) {
+        case YIN_ARG_NAME:
+            /* check for multiple definitions of name */
+            LY_CHECK_ERR_RET((*mod)->mod->name, LOGVAL_YANG(xml_ctx, LYVE_SYNTAX_YIN, "Duplicit definition of module name \"%s\"",
+                                                            (*mod)->mod->name), LY_EEXIST);
+
+            /* read module name */
+            if (xml_ctx->status != LYXML_ATTR_CONTENT) {
+                LOGVAL(xml_ctx->ctx, LY_VLOG_LINE, &xml_ctx->line, LYVE_SYNTAX, "Missing value of argument \"name\".");
+            }
+            ret = lyxml_get_string(xml_ctx, data, &buf, &buf_len, &out, &out_len, &dynamic);
+            LY_CHECK_ERR_RET(ret != LY_SUCCESS, LOGMEM(xml_ctx->ctx), LY_EMEM);
+            (*mod)->mod->name = lydict_insert(xml_ctx->ctx, out, out_len);
+            LY_CHECK_ERR_RET(!(*mod)->mod->name, LOGMEM(xml_ctx->ctx), LY_EMEM);
+            break;
+
+        default:
+            /* unrecognized attribute */
+            /* unrecognized or unexpected attribute */
+            /* TODO probably still can be from extension */
+            if (arg != YIN_ARG_UNKNOWN || name) {
```

I think that here we have to distinguish attributes from the yin namespace and from other namespaces. Unknown attributes in yin namespace are errors, but attributes from other namespaces are probably fine and should be silently ignored.